src/pybind/mgr/nfs/cluster.py

```diff
@@ -213,54 +218,58 @@ def show_nfs_cluster_info(self, cluster_id: Optional[str] = None) -> Tuple[int,
                 res = self._show_nfs_cluster_info(cluster_id)
                 if res:
                     info_res[cluster_id] = res
-            return (0, json.dumps(info_res, indent=4), '')
+            return info_res
         except Exception as e:
-            return exception_handler(e, "Failed to show info for cluster")
+            log.exception("Failed to show info for cluster")
+            raise ErrorResponse.wrap(e)
 
-    def get_nfs_cluster_config(self, cluster_id: str) -> Tuple[int, str, str]:
+    def get_nfs_cluster_config(self, cluster_id: str) -> str:
         try:
             if cluster_id in available_clusters(self.mgr):
                 rados_obj = self._rados(cluster_id)
                 conf = rados_obj.read_obj(user_conf_obj_name(cluster_id))
-                return 0, conf or "", ""
+                return conf or ""
             raise ClusterNotFound()
         except Exception as e:
-            return exception_handler(e, f"Fetching NFS-Ganesha Config failed for {cluster_id}")
+            log.exception(f"Fetching NFS-Ganesha Config failed for {cluster_id}")
+            raise ErrorResponse.wrap(e)
 
-    def set_nfs_cluster_config(self, cluster_id: str, nfs_config: str) -> Tuple[int, str, str]:
+    def set_nfs_cluster_config(self, cluster_id: str, nfs_config: str) -> None:
         try:
             if cluster_id in available_clusters(self.mgr):
                 rados_obj = self._rados(cluster_id)
                 if rados_obj.check_user_config():
-                    return 0, "", "NFS-Ganesha User Config already exists"
+                    raise ErrorResponse("NFS-Ganesha User Config already exists",
+                                        return_value=-errno.EEXIST)
```

Comment:
Here for instance: if automation tries to create a user, the manager or network restarts and the command gets resubmitted, and then the automation gets back an EEXIST error — that gets propagated to users as an error instead of a success message, and may result in broken state if higher-level layers needed to do more work to associate the new user with an existing tenant...

Reply:
OK, I think I understand where you are coming from. I think that if this was a new API we'd want to design it so it doesn't return an "error message" when a resource is being create-or-update'd but some sort of informational structured response. Example:
```
$ ceph foobar create user bob
{
  "user": "bob",
  "id": 509,
  "created": "2022-08-11T15:50+00Z"
}
```

And then that structured response would be returned regardless of whether "bob" was just created by this invocation or a year ago.  

For this existing API I can change the response code value back to zero. Then externally it should behave as it did previously.

I needed to write the above to "prove" to myself there's a way to cope with the scenario you outlined with these decorators.  My goal isn't to change everything just to make writing mgr modules more "pythonic" and abstract away some of the mgr-isms that seep into lower layers of the "business logic".

Reply:
> OK, I think I understand where you are coming from. I think that if this was a new API we'd want to design it so it doesn't return an "error message" when a resource is being create-or-update'd but some sort of informational structured response.


Ah yeah, that would make a lot of sense. Unfortunately not the way we went from our initially CLI-centric command "API" -- it was very much focused on users typing commands and the monitors responding to them. This might be a good longer-term project but replacing the existing duct tape without unpleasant behavior changes would be a pain.

Reply:
Idempotency behavior has been retained with the use of NonFatalError() that converts it to response tuple (0, msg, "").

Reply:
@phlogistonjohn I was manually testing this. Looks like the idempotent behavior of the command is not retained. 

```
$ ./bin/ceph nfs cluster config set nfs-ganesha -i ganesha.conf
$ ./bin/ceph nfs cluster config set nfs-ganesha -i ganesha.conf
Error EINVAL: NFS-Ganesha User Config already exists
$ echo $?
22
```
Same with other commands that raise NonFatalError(). I expected the command to return success. Unless I'm misunderstanding what NonFatalError() is supposed to do, I think there is a bug in NonFatalError() class.  

There are a two idempotency tests in qa/tasks/cephfs/test_nfs.py. I don't think they will succeed.
test_create_delete_cluster_idempotency
test_create_delete_export_idempotency

Reply:
OK, thanks for the heads up. I just did a build and am now set up to run teuthology on this pr.

Reply:
This is one of those, took me a while to see it but now it's obvious type issues. The problem is that we're trying to use specific exceptions but many of these functions catch all exceptions. Exmple:

```
try:
   x = do_stuff()
   if not x:
       raise NotFatalError("x not created")
   return
except Exception as e:
    raise ErrorResponse.wrap(e)
```

There are two choices that jump out at me, (1) we add new except clauses throughout the nfs py files like:
```
try:
   ....
except ErrorResponseBase:
    raise
except Exception as e:
    raise ErrorResponse.wrap(e)
```

(2) we teach `ErrorResponse.wrap(...)` not to change the error code for anything that already has one.

Even better I think would be to have wrap just return the `e` exactly as-is if it inherits from ErrorResponseBase. That's what I'll try first.  Give me a shout if you think otherwise - if it sounds good I'll update the patches, do another build, and test with that approach.

Reply:
Your commit https://github.com/ceph/ceph/pull/46209/commits/04d47f0535085894ced002437ab930d8f8b72b21 looks good. The approach makes sense.